portfolio/src/main/java/com/google/sps/servlets/DataServlet.java

```diff
@@ -14,41 +14,86 @@
 
 package com.google.sps.servlets;
 
+import com.google.sps.data.Comment;
+import com.google.appengine.api.datastore.DatastoreService;
+import com.google.appengine.api.datastore.DatastoreServiceFactory;
+import com.google.appengine.api.datastore.Entity;
+import com.google.appengine.api.datastore.PreparedQuery;
+import com.google.appengine.api.datastore.Query;
+import com.google.appengine.api.datastore.Query.SortDirection;
 import java.io.IOException;
 import javax.servlet.annotation.WebServlet;
 import javax.servlet.http.HttpServlet;
 import javax.servlet.http.HttpServletRequest;
 import javax.servlet.http.HttpServletResponse;
 import java.util.ArrayList;
+import java.time.LocalDate;
 import com.google.gson.Gson;
 
 /** Servlet that returns some example content. TODO: modify this file to handle comments data */
 @WebServlet("/data")
 public class DataServlet extends HttpServlet {
 
-    private ArrayList<String> messages;
-
     @Override
     public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
 
-        //Adds messages to the ArrayList:
-        messages = new ArrayList<String>();
-        messages.add("This is the first message.");
-        messages.add("This is the second message.");
-        messages.add("This is the third message.");
+        DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
+        Query query = new Query("Comment").addSort("date", SortDirection.DESCENDING);
+    
+        PreparedQuery results = datastore.prepare(query);
+
+        ArrayList<Comment> comments = new ArrayList<Comment>();
+        for(Entity entity : results.asIterable()){
+
+            String username = (String) entity.getProperty("username");
+            String message = (String) entity.getProperty("message");
+            String date = (String) entity.getProperty("date");
+
+            Comment current = new Comment(username, message, date);
+
+            comments.add(current);
 
-        String json = toJsonString(messages);
+        }
+
+        String json = toJsonString(comments);
 
         response.setContentType("application/json;");
         response.getWriter().println(json);
 
   }
 
+    @Override
+    public void doPost(HttpServletRequest request, HttpServletResponse response) throws IOException {
+
+        LocalDate messageDate = LocalDate.now();
+
+        String username = request.getParameter("username");
+        String message = request.getParameter("message");
+        String date = messageDate.toString();
+
+        if(username.length() > 1 && message.length() > 1){
```

Comment:
It's good that you do these checks – however it may be useful to return an error message to the user. It can be quite confusing for a user to hit enter on an empty comment and simply see the same page they saw before. For this, a page such as /error.html which contains a simple error message can be useful. [Don't feel like you have to implement it for this week 2 assignment, since this is beyond the scope of the assignment. Just something to keep in mind for the future.]

Reply:
I was thinking of returning a value inside the redirect like `response.sendRedirect("/index.html?comment_error")` and then check the URL within JS to display a pop-up message notifying the user about an invalid comment but didn't have time to implement it.